controller/volume_controller.go

```diff
@@ -1026,6 +1031,176 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
+	if isDataLocalityDisabled(v) {
+		return nil
+	}
+
+	if v.Status.State != types.VolumeStateAttached {
+		return nil
+	}
+
+	if e == nil {
+		return fmt.Errorf("BUG: ReconcileLocalReplica needs a valid engine")
+	}
+
+	if !vc.isSafeToModifyReplica(v, e, rs) {
+		return nil
+	}
+
+	if !hasLocalReplicaOnSameNodeAsEngine(e, rs) {
+		if _, err := vc.addLocalReplicaOnSameNodeAsEngine(v, e, rs); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// check if there are too many healthy replicas
+	healthyCount := 0
+	for _, r := range rs {
+		if r.Spec.FailedAt == "" && r.Spec.HealthyAt != "" {
+			healthyCount++
+		}
+	}
+
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
+
+	// we prefer to delete replicas on the same disk, then replicas on the same node, then replicas on the same zone
+	rNames, err := vc.getPreferredReplicaCandidatesForDeletion(rs)
+	if err != nil {
+		return err
+	}
+
+	// Randomly delete extra non-local healthy replicas in the preferred candidate list rNames
+	// Sometime reconcileLocalReplica() is called more than once with the same input (v,e,rs).
+	// To make the deleting operation idempotent and prevent deleting more replica than needed,
+	// we always delete the replica with the smallest name.
+	sort.Strings(rNames)
+	for _, rName := range rNames {
+		r := rs[rName]
+		if r.Spec.NodeID != e.Spec.NodeID {
+			return vc.deleteReplica(r, rs)
+		}
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) getPreferredReplicaCandidatesForDeletion(rs map[string]*longhorn.Replica) ([]string, error) {
+	diskToReplicaMap := make(map[string][]string)
+	nodeToReplicaMap := make(map[string][]string)
+	zoneToReplicaMap := make(map[string][]string)
+
+	nodeList, err := vc.ds.ListNodes()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range rs {
+		diskToReplicaMap[r.Spec.NodeID+r.Spec.DiskID] = append(diskToReplicaMap[r.Spec.NodeID+r.Spec.DiskID], r.Name)
+		nodeToReplicaMap[r.Spec.NodeID] = append(nodeToReplicaMap[r.Spec.NodeID], r.Name)
+		if node, ok := nodeList[r.Spec.NodeID]; ok {
+			zoneToReplicaMap[node.Status.Zone] = append(zoneToReplicaMap[node.Status.Zone], r.Name)
+		}
+	}
+
+	var deletionCandidates []string
+
+	// prefer to delete replicas on the same disk
+	deletionCandidates = findValueWithBiggestLength(diskToReplicaMap)
+	if len(deletionCandidates) > 1 {
+		return deletionCandidates, nil
+	}
+
+	// if all replicas are on different disks, prefer to delete replicas on the same node
+	deletionCandidates = findValueWithBiggestLength(nodeToReplicaMap)
+	if len(deletionCandidates) > 1 {
+		return deletionCandidates, nil
+	}
+
+	// if all replicas are on different nodes, prefer to delete replicas on the same zone
+	deletionCandidates = findValueWithBiggestLength(zoneToReplicaMap)
+	if len(deletionCandidates) > 1 {
+		return deletionCandidates, nil
+	}
+
+	// if all replicas are on different zones, return all replicas' names in the input RS
```

Comment:
```
if all replicas are on same zones
```

Reply:
I am not sure I understand this. Should it be `if all replicas are on different zones` because we checked for replicas in the same zone before come to this?

Reply:
Oh sorry for the misunderstanding... This should be good.